src/gsm/gsm.c

```diff
@@ -0,0 +1,241 @@
+ * this code. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "array_utils.h"
+#include "cellular.h"
+#include "gsm.h"
+#include "printk.h"
+#include "serial_buffer.h"
+
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+bool gsm_ping_modem(struct serial_buffer *sb)
+{
+        const char *msgs[2];
+        const size_t msgs_len = ARRAY_LEN(msgs);
+
+        serial_buffer_reset(sb);
+        serial_buffer_append(sb, "AT");
+        const size_t count =cellular_exec_cmd(sb, READ_TIMEOUT, msgs,
+                                              msgs_len);
+        return is_rsp_ok(msgs, count);
+}
+
+/**
+ * Turns on/off echo from the modem.
+ */
+bool gsm_set_echo(struct serial_buffer *sb, bool on)
+{
+        const char *cmd = on ? "ATE1" : "ATE0";
+        const char *msgs[2];
+        const size_t msgs_len = ARRAY_LEN(msgs);
+
+        serial_buffer_reset(sb);
+        serial_buffer_append(sb, cmd);
+        const size_t count = cellular_exec_cmd(sb, READ_TIMEOUT, msgs, msgs_len);
+        const bool status = is_rsp_ok(msgs, count);
+
+        if (!status)
+                pr_warning("[gsm] Failed to alter echoing.\r\n");
+
+        return status;
+}
+
+
+bool gsm_get_subscriber_number(struct serial_buffer *sb,
+                               struct cellular_info *ci)
+{
+        const char *cmd = "AT+CNUM";
+        const char *msgs[2];
+        const size_t msgs_len = ARRAY_LEN(msgs);
+
+        serial_buffer_reset(sb);
+        serial_buffer_append(sb, cmd);
+        const size_t count = cellular_exec_cmd(sb, MEDIUM_TIMEOUT, msgs,
+                                               msgs_len);
+        const bool status = is_rsp_ok(msgs, count);
+
+        if (!status) {
+                pr_warning("[gsm] Failed to read phone number\r\n");
+                return false;
+        }
+
+        char *num_start = strstr(msgs[0], ",\"");
+        if (!num_start)
+                goto parse_fail;
+
+        num_start += 2;
+        char *num_end = strstr(num_start, "\"");
+        if (!num_end)
+                goto parse_fail;
+
+        *num_end = '\0';
+        strncpy(ci->number, num_start, sizeof(ci->number));
+        return true;
+
+parse_fail:
+        pr_warning("[gsm] Failed to prase phone number\r\n");
+        return false;
+}
+
+bool gsm_get_signal_strength(struct serial_buffer *sb,
+                             struct cellular_info *ci)
+{
+        const char *cmd = "AT+CSQ";
+        char *msgs[2];
+        const char **const_msgs = (const char **) msgs; /* Stupid Compiler */
+        const size_t msgs_len = ARRAY_LEN(msgs);
+
+        serial_buffer_reset(sb);
+        serial_buffer_append(sb, cmd);
+        const size_t count = cellular_exec_cmd(
+                sb, READ_TIMEOUT, const_msgs, msgs_len);
+        const bool status = is_rsp_ok(const_msgs, count);
+
+        if (!status) {
+                pr_warning("[gsm] Failed to read signal strength\r\n");
```

Comment:
with these failure messages, would it be useful to also include the response received, for diagnostic purposes?